File: C/1935.c
#include <stdio.h>  // 후위 표기식2
#include <stdlib.h>
#include <string.h>

typedef struct
{
	double *stk;
	int ptr;
}STACK;

void push(STACK* s, double data)
{
	s->stk[s->ptr++] = data;
}

double pop(STACK* s)
{
	return s->stk[--s->ptr];
}

double calculate(char c, double first, double second)
{
	switch (c)
	{
	case '*':
		return first * second;
	case '/':
		return second / first;
	case '+':
		return first + second;
	case '-':
		return second - first;
	}
}


int main()
{
	int n;
	scanf("%d", &n);

	char str[101];
	scanf("%s", str);

	int arr[26];
	for (int i = 0; i < n; i++)
		scanf("%d", &arr[i]);

	STACK s;
	s.ptr = 0; s.stk = malloc(sizeof(double) * n);

	double first, second;
	for (int i = 0; i < strlen(str); i++)
	{
		if (str[i] >= 'A' && str[i] <= 'Z')			// 문자
		{
			push(&s, arr[str[i] - 'A']);
		}
		else			// 연산자
		{
			first = pop(&s);
			second = pop(&s);
			push(&s, calculate(str[i], first, second));
		}
	}

	double result = pop(&s);
	printf("%.2f\n", result);
	
	return 0;
}

/*
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct
{
	double* stk;
	int ptr;
}STACK;

int is_empty(STACK* s)
{
	if (s->ptr <= 0)
		return 1;
	else
		return 0;
}

void push(STACK* s, double data)
{
	s->stk[s->ptr++] = data;
}

double pop(STACK* s)
{
	if (!is_empty(s))
		return s->stk[--s->ptr];
}


int main()
{
	int n;
	scanf("%d", &n);

	char str[101];
	scanf("%s", str);

	double arr[26] = { 0, };
	for (int i = 0; i < n; i++)
		scanf("%lf", &arr[i]);

	STACK s;
	s.ptr = 0; s.stk = (double*)malloc(sizeof(double) * n);

	int len = strlen(str);
	double result = 0;
	for (int i = 0; i < len; i++)
	{
		if (str[i] >= 'A' && str[i] <= 'Z')
		{
			push(&s, arr[str[i] - 'A']);
		}
		else if (str[i] == '*')
		{
			double num1 = pop(&s);
			double num2 = pop(&s);
			result = num1 * num2;
			push(&s, result);
		}
		else if (str[i] == '/')
		{
			double num1 = pop(&s);
			double num2 = pop(&s);
			result = num2 / num1;
			push(&s, result);
		}
		else if (str[i] == '+')
		{
			double num1 = pop(&s);
			double num2 = pop(&s);
			result = num1 + num2;
			push(&s, result);
		}
		else if (str[i] == '-')
		{
			double num1 = pop(&s);
			double num2 = pop(&s);
			result = num2 - num1;
			push(&s, result);
		}
	}

	printf("%.2f", pop(&s));

	return 0;
}
*/

File: JAVA/src/_1935.java
import java.io.*;
import java.util.Stack;

// 후위 표기식2
public class _1935 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int n = Integer.parseInt(br.readLine());
        String input = br.readLine();
        Stack<Double> stack = new Stack<>();

        // arr에 피연산자 저장
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(br.readLine());
        }

        for (int i = 0; i < input.length(); i++) {
            char data = input.charAt(i);
            // 피연산자의 경우 --> stack에 push
            if (data >= 'A' && data <= 'Z') {
                stack.push((double) arr[data - 'A']);
            }
            // 연산자의 경우
            else {
                double num1 = stack.pop();
                double num2 = stack.pop();
                double res = 0.0;
                switch (data) {
                    case '*':
                        res = num1 * num2;
                        break;
                    case '/':
                        res = num2 / num1;
                        break;
                    case '+':
                        res = num1 + num2;
                        break;
                    case '-':
                        res = num2 - num1;
                }
                stack.push(res);
            }
        }

        String ans = String.format("%.2f", stack.pop());
        bw.write(ans);

        bw.flush();
        bw.close();

    }
}